### ISSUE TITLE
Different behavior from mapping an item or List of items

### ISSUE THREAD

**Issue by lucax88x** [AutoMapper/AutoMapper#1135]
When copying from a dto to a object, and dto misses some properties, the object will keep his values.

It behaves correctly when copying the item itself, or even a wrapper of the item itself, but when we have the dto and the object in lists, it fails.

Reproducible code: https://dotnetfiddle.net/Qyruqn

**Comment by jbogard** [comment]
This is the expected behavior. When mapping between existing objects, AutoMapper will map all properties. So ID is not pulled over.

When mapping collections, AutoMapper has no idea on how to merge collections. The default behavior is to clear the destination collection and create new mapped objects for the source items.

Try writing this code without AutoMapper, just manually coding it. You'll find it's not easy to merge objects, which is what you're trying to do. Suppose there are two items in the destination list, but the order is different than the DTOs? Your DTOs don't have IDs, so how would AutoMapper know how to line things up? What if items are added, deleted, changed or unchanged in the source collection? Should AutoMapper remove items in the destination collection or just ignore them?

In short, AutoMapper is not an ORM, which has all these complicated rules for translating collections to SQL records. I don't really want it to be an ORM, so AutoMapper won't support these sort of complicated rules.

**Comment by TylerCarlson1** [comment]
https://github.com/TylerCarlson1/Automapper.Collection does this.  But every thing @jbogard says is true.  You have to provide Automapper.Collection some way to compare the dto to object, so it knows how to add,update,delete.  There's defaults for it, but if you don't have ID's or any other unique identifiers this is impossible

**Comment by lucax88x** [comment]
Agree with you both, I did not expect AutoMapper to understand how to modify the list (add, update, delete).

I expected that AutoMapper, when I provide a List of dtos and a list of entities, just to correctly maps it one by one, without matching it by id, etc.

It's my job then to provide in both lists the same order, same count, etc.

Now I go back to my point.

When I provide a list of ONE dto, and a list of ONE entity, I expect him to copy properties as it's copying the items without a list.

What I've encountered and I've reported in the example in the first post, it's that AutoMapper is trying to copy the properties nicely, but has just one flaw, if a source property is null or empty it copy to the destination property. 

stupid example:

AutoMapper.Map(dto, entity)
AutoMapper.Map(new []{ dto }, new []{ entity })

This, should behave, for me, the SAME way.

I'm sure you may have encountered a lot of issues on trying to what I'm saying, so if you will say it's not an issue, it will be ok for me. I'll proceed on how I'm doing it right now, just doing a foreach in the list and applying the AutoMapper.Map on the single items.

**Comment by jbogard** [comment]
You're in the minority there. Most expect to somehow line up the lists and
intelligently merge items. It would be more confusing to make an assumption
that both lists are the same size and represent the same items, but throw
an exception otherwise.

So I won't make any assumptions about the destination collections, and
leave any more complex scenario to manual mapping, which is far less
magical and far more obvious.
